src/middleware/bodyParser/formData.ts

```diff
@@ -0,0 +1,104 @@
+/*!
+ * Port of body-parser (https://github.com/expressjs/body-parser) for Deno.
+ *
+ * Licensed as follows:
+ *
+ * (The MIT License)
+ *
+ * Copyright (c) 2014 Jonathan Ong <me@jongleberry.com>
+ * Copyright (c) 2014-2015 Douglas Christopher Wilson <doug@somethingdoug.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * 'Software'), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+import type { NextFunction, OpineRequest, OpineResponse } from "../../types.ts";
+import { hasBody, MultipartReader } from "../../../deps.ts";
+import { typeChecker } from "./typeChecker.ts";
+
+/**
+ * Create a middleware to parse formData.
+ *
+ * @param {object} [options]
+ * @return {function}
+ * @public
+ */
+export function formData(options: any = {}) {
+  const type = options.type || "multipart/form-data";
+  const verify = options.verify || false;
+
+  if (verify !== false && typeof verify !== "function") {
+    throw new TypeError("option verify must be function");
+  }
+
+  // create the appropriate type checking function
+  const shouldParse = typeof type !== "function" ? typeChecker(type) : type;
+
+  return function formParser(
+    req: OpineRequest,
+    _res: OpineResponse,
+    next: NextFunction,
+  ) {
+    if (req._parsedBody) {
+      next();
+
+      return;
+    }
+
+    // skip requests without bodies
+    if (
+      !hasBody(req.headers) ||
+      parseInt(req.headers.get("content-length") || "") === 0
+    ) {
+      req.parsedBody = new FormData();
+      req.body = req.parsedBody;
+
+      next();
+
+      return;
+    }
+
+    // determine if request should be parsed
+    if (!shouldParse(req)) {
+      next();
+
+      return;
+    }
+
+    // read
+    const type = req.headers.get("Content-Type");
+
+    if (!type) {
+      req.parsedBody = new FormData();
+      req.body = req.parsedBody;
+
+      next();
+
+      return;
+    }
+    const mr = new MultipartReader(req.raw, type.split(";")[1]?.split("=")[1]);
```

Comment:
Slightly uneasy about the boundary retrieval code here, though haven’t found anything to suggest that this wouldn’t work just fine given doesn’t look to be any other header params other than the boundary. Nevertheless would be good to not assume users always do sensible things - consider implementing a header param lookup which is safeguarded against there being other non-boundary params like oakserver has https://github.com/oakserver/oak/blob/main/multipart.ts#L13

Reply:
Well yeah, mostly. I'm not entierly sure if I still have to have the copyright there because I did copy from json.ts to formData.ts and just changed stuff around 🤷🏽  Gonna get rid of it for now since I actually wrote the rest of the middleware myself

Reply:
would something work like `boundary=(.*?)(;|$)` it'd just pick up the first boundary found in content type?